### ISSUE TITLE
Make an op compute its input once and hold.

### ISSUE THREAD

**Issue by jonasrauber** [tensorflow/tensorflow#6399]
Currently, tf.constant can take NumPy arrays as input, but it cannot take TensorFlow tensors as input. It would be useful to allow tf.constant to take tensors (that change with every evaluation, e.g. the tensor created by a dequeue op) and turn it into a constant tensor by evaluating the input tensor only once.

**Comment by yaroslavvb** [comment]
Constants are special in that their value is inlined into the graph, so the value is defined before session is created. Functionality like static shape-inference relies on this to figure out shapes before the first run call is created. If you allow constants to be undefined until the first session run, that would complicate that logic.

You could implement the functionality you want by using variables. Use `collections=[]` to exempt the variable from `tf.initialize_all_variables` and run the initializer manually

```
queue1 = tf.FIFOQueue(capacity=10, dtypes=[tf.int32], shapes=[()])
sess = tf.Session()
sess.run(queue1.enqueue(1))
a = tf.Variable(queue1.dequeue(), collections=[])
sess.run(a.initializer)

```

**Comment by aselle** [comment]
@yaroslavvb's comment seems to resolve this issue. please comment if this does not resolve your question @jonasrauber . Thanks for the issue.

**Comment by jonasrauber** [comment]
Kind of, but not really. I am aware that it's possible to use workarounds like the one proposed by @yaroslavvb. I don't care so much how it relates to real constants, I just thought it would be nice to have an op that consumes it's input once and then outputs this initial input whenever it's evaluated. A possible use case would be to turn a data loading queue into a fake data queue, to measure network performance independent of IO performance, etc. (basically like tf.zeros_like or tf.ones_like, just with a more natural data distribution based on a single example from the queue). If you don't consider this important enough, I am fine with closing this issue and using workarounds.

**Comment by aselle** [comment]
@jonasrauber: I changed the title, because essentially, I don't think what your describing could be a constant.  I would be worried that such an op would be somewhat confusing. What if upstream values had been computed with some temporary feed and then the feed was changed. The first value evaluated was not useful, so you don't want to keep that cached value, you want the second input value. @mrry, @martinwicke, do you have any thoughts on the semantics of this operation?

**Comment by mrry** [comment]
Off the top of my head, there are a few different possibilities here (although I'm sure there are others...):

* The easiest would be to implement a write-once version of `tf.Variable`. It would still have an explicit `initialize` op, but it could be treated as a constant subsequently (e.g. in graph optimizations constant folding, CSE, etc.). Depending on how you want to use it, it could have blocking read semantics (block until initialized) or it could raise an error (like a normal `tf.Variable`). With the blocking read semantics, it would essentially be an [I-Structure](http://www.cse.chalmers.se/edu/year/2016/course/pfp/Papers/IStructures.pdf) (see Section 5 for details), but it would also be another potential source of deadlock :).

* You could build a constant-like construct using a TensorFlow function (or perhaps a `tf.Variable` and `tf.cond()`, if you're not worried about concurrent assignments). On the first execution, it would call the function and store the result in the "constant", then produce that value. Subsequent executions would produce the same value. We discussed doing something like this for `tf.Variable` initialization a long time ago, but neither functions nor control flow were stable enough at the time.

TensorFlow's push semantics make it tricky to implement a construct that takes a `tf.Tensor` and conditionally executes the subgraph needed to produce that tensor. I'm sure it could be done, but it would require changes in the core executor, so either of the above two options would be preferable.

**Comment by yaroslavvb** [comment]
BTW, regarding recipe 2, I've been using `cond` to initialize variables once as [here](https://gist.github.com/yaroslavvb/d592394c0cedd32513f8fbb87ca05938#file-smart_initialize-py-L22). Without `tf.cond` you get unexpected results when initialization is split over several `sess.run` calls (https://github.com/tensorflow/tensorflow/issues/4920)

**Comment by mrry** [comment]
@yaroslavvb Right, that code should work, as long as you're okay with the possibility that the initializer might run more than once.

**Comment by aselle** [comment]
Closing due to lack of recent activity. if you provide additional information, we can reopen this. Thanks!

**Comment by yunjiangster** [comment]
> BTW, regarding recipe 2, I've been using `cond` to initialize variables once as [here](https://gist.github.com/yaroslavvb/d592394c0cedd32513f8fbb87ca05938#file-smart_initialize-py-L22). Without `tf.cond` you get unexpected results when initialization is split over several `sess.run` calls (#4920)

Thanks for the nice tip. I tried adding the final assign op of an initialization function to the TABLE_INITIALIZERS collection before, but only recently realized that it still computes the whole function every time a request comes in during serving. Could you help me understand why it behaves this way?
